src/Compilers/CSharp/Test/Semantic/Semantics/NullCheckedParameterTests.cs

```diff
@@ -518,5 +529,128 @@ public void M()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(out string x!)
+    {
+        x = ""hello world"";
+    }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(
+                    // (5,31): error CS8720: By-reference parameter 'x' cannot be null-checked.
+                    //     public void M(out string x!)
+                    Diagnostic(ErrorCode.ERR_NullCheckingOnByRefParameter, "!").WithArguments("x").WithLocation(5, 31));
+        }
+
+        [Fact]
+        public void TestNullCheckedRefString()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(ref string x!)
+    {
+        x = ""hello world"";
+    }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(
+                    // (5,31): error CS8720: By-reference parameter 'x' cannot be null-checked.
+                    //     public void M(ref string x!)
+                    Diagnostic(ErrorCode.ERR_NullCheckingOnByRefParameter, "!").WithArguments("x").WithLocation(5, 31));
+        }
+
+        [Fact]
+        public void TestNullCheckedInString()
+        {
+            var source = @"
+class C
+{
+    public static void Main() { }
+    public void M(in string x!) { }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(
+                    // (5,30): error CS8720: By-reference parameter 'x' cannot be null-checked.
+                    //     public void M(in string x!) { }
+                    Diagnostic(ErrorCode.ERR_NullCheckingOnByRefParameter, "!").WithArguments("x").WithLocation(5, 30));
+        }
+
+        [Fact]
+        public void TestNullCheckedGenericWithDefault()
+        {
+            var source = @"
+class C
+{
+    
+    static void M1<T>(T t! = default) { }
+    static void M2<T>(T? t! = default) where T : struct { }
+    static void M3<T>(T t! = default) where T : class { }
+}";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(
+                    // (6,26): warning CS8721: Nullable value type 'T?' is null-checked and will throw if null.
+                    //     static void M2<T>(T? t! = default) where T : struct { }
+                    Diagnostic(ErrorCode.WRN_NullCheckingOnNullableValueType, "t").WithArguments("T?").WithLocation(6, 26),
+                    // (7,30): warning CS8719: Parameter 'T' is null-checked but is null by default.
+                    //     static void M3<T>(T t! = default) where T : class { }
+                    Diagnostic(ErrorCode.WRN_NullCheckedHasDefaultNull, "default").WithArguments("T").WithLocation(7, 30));
+        }
+
+        [Fact]
+        public void TestNullableInteraction()
+        {
+            // PROTOTYPE : Add warning about combining explicit null checking with a nullable
```

Comment:
>Add warning about combining explicit null checking with a nullable [](start = 27, length = 66)

What are we expecting to test: `object? s!`?

Reply:
It was just `int? x!`, will remove.